perf: sonic boom

H4ad opened the pull request:
Hello again!

This time I bring more performance improvements based on my study of PinoJS, we went from 102k in this https://github.com/jmcdo29/ogma/pull/1491 to:

```
'NestJS Logger x 6,598 ops/sec ±5.64% (80 runs sampled)',
'Ogma x 419,803 ops/sec ±5.68% (76 runs sampled)',
'Pino x 743,735 ops/sec ±2.15% (82 runs sampled)',
'console.log x 140,086 ops/sec ±1.25% (89 runs sampled)',
'process.stdout x 288,950 ops/sec ±2.77% (75 runs sampled)'
```

Now, the profiler information is almost empty hahaha:

```
Statistical profiling result from perf-ogma-stress-v4.log, (4072 ticks, 0 unaccounted, 0 excluded).

 [Shared libraries]:
   ticks  total  nonlib   name
   1659   40.7%          /home/h4ad/.asdf/installs/nodejs/16.16.0/bin/node
    408   10.0%          /usr/lib/x86_64-linux-gnu/libc-2.31.so
     30    0.7%          [vdso]

 [JavaScript]:
   ticks  total  nonlib   name
     44    1.1%    2.2%  LazyCompile: *formatStream /home/h4ad/Projects/opensource/performance-analysis/ogma-performance-analysis/node_modules/@ogma/logger/src/logger/ogma.js:168:17
     26    0.6%    1.3%  LazyCompile: *<anonymous> /home/h4ad/Projects/opensource/performance-analysis/ogma-performance-analysis/dist/ogma-log.stress.js:1:1
      3    0.1%    0.2%  Function: ^SonicBoom.release /home/h4ad/Projects/opensource/performance-analysis/ogma-performance-analysis/node_modules/sonic-boom/index.js:127:18
      1    0.0%    0.1%  LazyCompile: *write node:fs:802:15
      1    0.0%    0.1%  LazyCompile: *normalizeString node:path:66:25
      1    0.0%    0.1%  LazyCompile: *info /home/h4ad/Projects/opensource/performance-analysis/ogma-performance-analysis/node_modules/@ogma/logger/src/logger/ogma.js:238:9
      1    0.0%    0.1%  LazyCompile: *SonicBoom.release /home/h4ad/Projects/opensource/performance-analysis/ogma-performance-analysis/node_modules/sonic-boom/index.js:127:18
      1    0.0%    0.1%  Function: ^wrapper node:fs:803:19
      1    0.0%    0.1%  Function: ^normalizeString node:path:66:25
```

> The number of ticks is bigger, but the time to perform the logs is now much smaller than in the previous version.

> [Source](https://github.com/jmcdo29/ogma/files/10388229/perf-ogma-stress-v4.txt)

The first optimization was the usage of [sonic-boom](https://www.npmjs.com/package/sonic-boom), this library was responsible for more of the performance improvements.

Then, I did a breaking change in `styler` by moving from `getters` to use methods because is way faster.

Finally, the last optimization was inside `ogma.ts`, I almost cache all calls for `colorize`, we had a bunch of places where the value didn't change so we don't need to calculate every time we log something and I also did a happy path in the code to single messages to avoid perform `regex.test`.

H4ad committed:
perf(styler): prefer methods over getters to faster call

jmcdo29 committed:
chore: benchmark with sonic boom

jmcdo29 committed:
feat: directly assign most of json log object

jmcdo29 committed:
feat: fix up the hidden class and cli types

H4ad commented:
With the latest changes, the benchmark looks like:

```
  'NestJS Logger x 6,797 ops/sec ±5.41% (80 runs sampled)',
  'Ogma x 497,586 ops/sec ±4.41% (78 runs sampled)',
  'Pino x 741,796 ops/sec ±1.29% (88 runs sampled)',
  'console.log x 136,191 ops/sec ±1.83% (91 runs sampled)',
  'process.stdout x 294,599 ops/sec ±1.52% (67 runs sampled)'

  'NestJS Logger x 6,816 ops/sec ±5.62% (83 runs sampled)',
  'Ogma x 500,222 ops/sec ±6.06% (66 runs sampled)',
  'Pino x 719,338 ops/sec ±2.24% (82 runs sampled)',
  'console.log x 134,167 ops/sec ±2.01% (93 runs sampled)',
  'process.stdout x 304,896 ops/sec ±2.18% (71 runs sampled)'
```

jmcdo29 committed:
chore: remove array.join

jmcdo29 committed:
fix: revert circular replacer change that broke tests

jmcdo29 commented:
I'm gonna play around with the benchmarks and improvements just a little more before merging this in. Trying to get the JSON logger up to the same kind of speed. Thanks for getting this started though! Definitely been a fun learning process

H4ad committed:
perf(logger): cache computed values and improve lookups for circular function

H4ad committed:
perf(logger): perform string concatenation of hot path instead json stringify

H4ad committed:
perf(logger): perform string concatenation of hot path instead json stringify

jmcdo29 commented on packages/logger/src/logger/ogma.ts:
```diff
@@ -114,102 +183,149 @@ export class Ogma {
+    if (this.options.levelKey) fastJson += `"${this.options.levelKey}":${mappedLevel},`;
+
+    if (typeof message === 'object')
+      fastJson += `"message":${stringify(message, this.circularReplacer())}`;
```
What happens if we `stringify` something that isn't an object? Does `fast-safe-stringify` care at all?

H4ad replied:
I think it will have the same behavior as `JSON.stringify`, for some primitives it will handle but for things like `Symbol` or `BigInt`, it brokes or return `undefined`.

jmcdo29 replied:
My thought here is that we already handle the cases of `Symbol`, `Function`, and `BigInt` in the circular replacer, so could we just use `fastJson += "message":${stringify(message, this.circularReplacer()}`; and not worry about the if check here? Not sure it would change the performance, but it's worth the though.

however, if we do that, I think we end up wrapping non-string values in quotes, which may not be what's desired here, so ignore the thought.

jmcdo29 commented on packages/logger/src/logger/ogma.ts:
```diff
@@ -14,8 +15,31 @@ const checkIfHasSpaceRegex = /[^\n]$/;
+    ['number', true],
+    ['string', true],
+    ['boolean', true],
+    ['number', true],
```
Is `number` supposed to be here twice?

H4ad replied:
Nop, fixed!

jmcdo29 commented:
I'm gonna make a couple of changesets for this to document all of the changes in an easy to digest manner. Once a PR for the nest core repo gets merged in with some extra websocket support I can release the next version of `@ogma/nestjs-module` and then I can make all of the breaking changes I'm going to at once. There's a lot of them in the pipeline now and I'd rather just do it all in one fell swoop, which is why I'm keeping this open for now. Don't worry though, it is not forgotten :smile: :rocket:

jmcdo29 committed:
chore(common): add changeset for time as number

jmcdo29 committed:
chore(cli): handle time varaible as number and string

jmcdo29 commented on packages/logger/test/ogma.spec.ts:
```diff
@@ -143,7 +143,7 @@ OgmaSuite(
     match(loggedVal, /\[Circular\]/);
     match(loggedVal, /\[Function:/);
-    match(loggedVal, /\[Symbol\(hello\)\]/);
+    match(loggedVal, 'hello');
```
What caused this to no longer print out `[Symbol(hello)]`?

jmcdo29 replied:
Doing some debugging, looks like something isn't quite what we're expecting right here:
![image](https://user-images.githubusercontent.com/28268680/213585069-4e4298bc-546b-4e89-9a49-1b3df1dcdab2.png)

Unless this is a part of the breaking changes. `symbol.toString()` already makes the value be `Symbol(value)` so we may want to revisit what the circular replacer is returning for symbols

H4ad replied:
We can change to `symbol.description` instead, so it will not duplicate the value `Symbol`.

jmcdo29 replied:
With how I originally had it written, logging out, `Symbol(value)`, I was mimicking the behavior of `console.log` (while adding the brackets around it). Now, we'd be making our own format, but being consistent within the logger itself as `BigInt` and `Function` already use this `key: value` syntax :thinking:

Yeah, I think let's change it to `symbol.description` and note it in the breaking changes

jmcdo29 replied:
Let's also update the test to show the full syntax being checked for, for the sake of verification and security in future updates

jmcdo29 committed:
chore(styler): add changeset for style methods

jmcdo29 committed:
feat: use symbol deescription for log uniformity

Functions and BigInt already use Function: thing as their format
so now symbol does too instad of using Symbol(thing)

jmcdo29 committed:
chore: update styler and logger changeset

jmcdo29 committed:
chore: clean up benchmark output

jmcdo29 committed:
docs: update styler chained methods

jmcdo29 commented:
Okay. I _think_ I'm done modifying this PR until it's time to merge it. Thanks for all your work again @H4ad! Really excited about the updates to this

jmcdo29 commented:
Oh boy the merge conflicts here are gonna be _fun_

H4ad committed:
perf(logger): added sonic boom instead writing directly to process.stdout

H4ad committed:
perf(logger): cache computed values & happy path for non-array messages

H4ad committed:
feat: squeeze out a bit more performance

H4ad committed:
feat: benchmarks using benchark fully set up

H4ad committed:
chore: update styler and nestjs-module changeset

H4ad committed:
chore: update lockfile

H4ad committed:
chore(logger): removed unused console log

H4ad commented:
@jmcdo29  Solved :)

jmcdo29 commented:
Oh, dude, thank you so much
